File: Data Structures & Algorithms/sqrtx/submission-0.py
class Solution:
    def mySqrt(self, x: int) -> int:

        start, end = 0, x
        result = 0

        

        while start <= end:
            mid = start + (end - start) // 2

            if mid * mid == x:
                return mid
            elif mid * mid < x:
                start = mid + 1
                result = mid 
            else:
                end = mid - 1
        
        return result

File: Data Structures & Algorithms/search-insert-position/submission-0.java
class Solution {
    public int binarySearch(int nums[], int target, int start, int end) {
        if (start > end)
            return start;
        
        int mid = start + (end - start) / 2;

        if (target == nums[mid]) {
            return mid;
        }

        if (target < nums[mid])
            return binarySearch(nums, target, start, mid - 1);
        return binarySearch(nums, target, mid + 1, end);
    }

    public int searchInsert(int[] nums, int target) {
        return binarySearch(nums, target, 0, nums.length - 1);
    }
}